src/libfsm/walk2.c

```diff
@@ -0,0 +1,469 @@
+#include "walk2.h"
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <errno.h>
+
+#include <adt/set.h>
+
+#include <fsm/fsm.h>
+#include <fsm/out.h>
+#include <fsm/options.h>
+#include <fsm/bool.h>
+#include <fsm/pred.h>
+#include <fsm/walk.h>
+
+#include "internal.h"
+
+/* XXX - revisit what would be a good size for this. */
+enum { FSM_WALK2_TUPLE_POOL_SIZE = 1024 };
+
+/* comparison of fsm_walk2_tuples for the (ordered) set */
+static int
+cmp_walk2_tuple(const void *a, const void *b)
+{
+	const struct fsm_walk2_tuple *pa = a, *pb = b;
+	ptrdiff_t delta;
+
+        /* XXX - do we need to specially handle NULLs? */
+
+	delta = (pa->a > pb->a) - (pa->a < pb->a);
+	if (delta == 0) {
+		delta = (pa->b > pb->b) - (pa->b < pb->b);
+	}
+
+	return delta;
+}
+
+struct fsm_walk2_tuple_pool {
+	struct fsm_walk2_tuple_pool *next;
+	struct fsm_walk2_tuple items[FSM_WALK2_TUPLE_POOL_SIZE];
+};
+
+void
+fsm_walk2_data_free(struct fsm_walk2_data *data)
+{
+	struct fsm_walk2_tuple_pool *p, *next;
+
+	if (data->states) {
+		set_free(data->states);
+	}
+
+	if (data->new) {
+		fsm_free(data->new);
+	}
+
+	for (p = data->head; p != NULL; p = next) {
+		next = p->next;
+		free(p);
+	}
+}
+
+static struct fsm_walk2_tuple *
+alloc_walk2_tuple(struct fsm_walk2_data *data)
+{
+	static const struct fsm_walk2_tuple zero;
+
+	struct fsm_walk2_tuple *item;
+	struct fsm_walk2_tuple_pool *pool;
+
+	if (data->head == NULL) {
+		goto new_pool;
+	}
+
+	if (data->top >= (sizeof data->head->items / sizeof data->head->items[0])) {
+		goto new_pool;
+	}
+
+	item = &data->head->items[data->top++];
+	*item = zero;
+	return item;
+
+new_pool:
+	pool = malloc(sizeof *pool);
+	if (pool == NULL) {
+		return NULL;
+	}
+
+	pool->next = data->head;
+	data->head = pool;
+	data->top = 1;
+
+	item = &pool->items[0];
+	*item = zero;
+	return item;
+}
+
+static unsigned
+walk2mask(int has_a, int has_b)
+{
+	int endbit = (!!has_a << 1) | !!has_b;
```

Comment:
Is this reproducing the same bits as the enum provides?

Reply:
Okay. Then I don't like that the knowledge about bits is repeated in those two places. Could you phrase this in terms of the enum's constants perhaps?

Reply:
I can't think of a way to do this without a lookup table, several ?: expressions, or a bunch of if statements.  Seems like a waste to introduce a 2x2 lookup table here, and the other options add some ugly branching.

Reply:
I was picturing if statements. or `switch ((has_a << 1) | has_b)`. The point being that that's in terms of the local variables here, and doesn't make any assumptions about the bit patterns for the enum constants.